src/main/java/jenkins/plugins/openstack/compute/JCloudsPreCreationThread.java

```diff
@@ -0,0 +1,92 @@
+package jenkins.plugins.openstack.compute;
+
+import java.lang.Math;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import hudson.Extension;
+import hudson.model.TaskListener;
+import hudson.model.AsyncPeriodicWork;
+
+/**
+ * Periodically ensure enough slaves are created.
+ */
+@Extension @Restricted(NoExternalUse.class)
+public final class JCloudsPreCreationThread extends AsyncPeriodicWork {
+    private static final Logger LOGGER = Logger.getLogger(JCloudsPreCreationThread.class.getName());
+
+    public JCloudsPreCreationThread() {
+        super("OpenStack slave pre-creation");
+    }
+
+    @Override
+    public long getRecurrencePeriod() {
+        return MIN * 2;
+    }
+
+    @Override
+    public void execute(TaskListener listener) {
```

Comment:
I think this deserves more documentation - either on code level or user level - to clarify how is this expected to work.

- As far as I understand the implementation, when configured on cloud level, all templates would simply inherit the number and keep X instances up for every template.
- This implementation seems to permit bypassing the instance capacity limit. Consider following: both cloud and template have `instanceCap == 3` and `instancesMin == 2`, there is 2 busy nodes with `retentionTime == 0` running. This gives us `maxNodes == 2`, `runningNodeTotal == 0` (as they are used / will not be reused) and `desiredNewInstances == 2`. Once the provisioning kicks in, there will be 4 nodes exceeding the instance capacity of 3.
- That reminds me, how will this go along with retention time, especially short one? I suspect those machines will be reprovisioned every x minutes even when there is no load as Jenkins consider the node to be idle since the launch if nothing run there. This was not a problem until now as that seldom ever happened and if, it was because the load has vanished so the node was cleaned up nicely. IMO, the machines should be up without reprovisioning until actually utilized.

Reply:
Documentation: sure, I will add a bit of documentation in the JCloudsPreCreationThread.java (which contains the main class).

Bypassing instance capacity: that's correct; this was a deliberate choice but I'm happy to revise the behaviour if you think it's better.  In your example, as you point out, the number of instances will exceed the capacity limit for a brief period of time (at most the retention time); but note that it will be only a temporary situation until the instances are disposed of. My thinking was that it was a bit better to go over the capacity limit but to be 100% sure that new instances are ready at all times.  If you think the capacity limit should be the deciding factor, I'm happy to change this.

Reprovisioning: you're right; again it's a choice: do we want the retention time to be strictly enforced or are we okay with machines sticking around for more than the retention time if they are not used? I've been a bit caution for the first implementation since the original code had a very fragile way to detect if a machine had been used or not. Now that we have something more solid here I agree with you that keeping unused machines even if retention time conflicts with this is the way to go. I will change this.

Reply:
> That reminds me, how will this go along with retention time, especially short one?[...]

Actually, scrap what I said earlier: if the retention time is > 0, the code in shouldSlaveBeRetained simply considers old and new machines to be potential candidates to be kept around.  As a result Jenkins will simply maintain the expected number of slaves, without constantly reprovisioning.

Reply:
> My thinking was that it was a bit better to go over the capacity limit but to be 100% sure that new instances are ready at all times. If you think the capacity limit should be the deciding factor, I'm happy to change this.

I prefer this to be the other way around. The capacity serves as a gatekeeper to prevent Jenkins to utilize whole quota which the plugin does not handle very well and can impact other users of the OS tenant.

> do we want the retention time to be strictly enforced or are we okay with machines sticking around for more than the retention time if they are not used?

I do not think sticking to the retention time works with the feature implemented here so I agree to change that, too. IOW, retention time should be enforced only for used nodes or for those its removal would not violate the `instancesMin` configuration.

Reply:
Capacity limit and retention time: agreed, this is what is implemented in the latest code.